### PR TITLE
Plans (grid): Disable term-savings price display if another discount is active (intro/coupon/etc.)

### PR DESCRIPTION
**Pull request by chriskmnds**
Related to https://github.com/Automattic/martech/issues/3403
Related to https://github.com/Automattic/wp-calypso/pull/96357
Branched from https://github.com/Automattic/wp-calypso/pull/96492

## Proposed Changes

This is a follow-up to the term-savings price display (see core implementation: https://github.com/Automattic/wp-calypso/pull/96357) to disable the term-savings display if any plan, across any available/effective term for the rendered grid appears as discounted already.

See related discussion: pdvytD-Se-p2#comment-645. Implementation directly addresses comment pdvytD-Se-p2#comment-650:
- a plan on a "one time discount" coupon will disable the term-savings
- a plan on a "special offer" introductory offer will disable the term-savings
- term-savings are disabled for all the plans, so switching between monthly/yearly/2yearly/3yearly will not resurface the term-savings, as long as any visible plan is discounted

## Why are these changes being made?

Avoid confusion when plans are discounted while the term-savings display is active. Keep the focus on the effective/promoted discount across the user's interaction with the grid.

See related discussion: pdvytD-Se-p2#comment-645. Implementation directly addresses comment pdvytD-Se-p2#comment-650

## Testing Instructions

TBD - WIP

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by chriskmnds** [comment]
@jeyip This is a pretty insane change. If this is where we want things to head, then I'll continue pushing through more surrounding refactors. I think this is where we want things to head, conceptually.

cc @southp as something to keep an eye on. The primary consideration going forward should on maintaining the memoization of the affected hooks. Otherwise, it literally can be 4x the processing on any interaction (if not already - needs to be tested). 🤷

**Comment by chriskmnds** [comment]
> @jeyip This is a pretty insane change. If this is where we want things to head, then I'll continue pushing through more surrounding refactors. I think this is where we want things to head, conceptually.
> 
> cc @southp as something to keep an eye on. The primary consideration going forward should on maintaining the memoization of the affected hooks. Otherwise, it literally can be 4x the processing on any interaction (if not already - needs to be tested). 🤷

Well, I'm going to experiment with another approach. This feels a little over the line.

**Comment by jeyip** [comment]
> @jeyip This is a pretty insane change. If this is where we want things to head, then I'll continue pushing through more surrounding refactors. I think this is where we want things to head, conceptually. 

> > Well, I'm going to experiment with another approach. This feels a little over the line.

Thanks for exploring this change. Pretty unfortunate that it gets crazy. I haven't reviewed the code yet, but I'll take a closer look before I sign off

**Comment by jeyip** [comment]
While I recognize that we might abandon this PR, I rebased this branch with `trunk` and resolved merge conflicts to better understand the breadth and width of what we're updating. There were about 15 extra files conflated with our code changes, which have since been resolved and removed. 

Things should be easier to wrap my head around now ( and also for anyone else that's following along )

I smoke tested behavior and it looks correct. 🙂

**Review by jeyip — COMMENTED** [review]
Ah. After digging into the code a bit more, I understand the concern. Summarizing to make sure I understand correctly, but feel free to chime in with any important info I'm missing.

Whereas previously we were retrieving and joining plans data only for the term being rendered ( monthly, _or_ annual, _or_ 2year, etc. ), we are now doing so for _all_ billing terms in the `useGridPlans` hooks. Besides the added code complexity, we're also seeing O(n) functions become O(n^2) [here](https://github.com/Automattic/wp-calypso/pull/96631/files#diff-43d0e51c52cc2961e8d8244ae1416f67f45e211c275961d241a775a9c4f0c5c0R100-R103) and [here](https://github.com/Automattic/wp-calypso/pull/96631/files#diff-18214fd84316736ff6662f03fa201b6255f4c20f277c0eb683b5e36e9c9dea6dR760-R762)

We're doing so because we'd like to calculate whether or not special offers, coupons, or discounts are visible to _any_ plan on _any_ term, at which point we can decide whether or not we want to render term savings.

I'll start thinking about alternate approaches as well.

File: client/my-sites/plans-features-main/index.tsx [hunk 1]
```diff
@@ -754,6 +757,52 @@ const PlansFeaturesMain = ( {
...
+			( isEnabled( 'plans/term-savings-price-display' ) ||
+				longerPlanTermDefaultExperiment.isEligibleForTermSavings ) &&
+			isInSignup
+		);
+	}, [
+		pricingForAllDisplayedIntervals,
+		longerPlanTermDefaultExperiment.isEligibleForTermSavings,
+		isInSignup,
+	] );
+
```
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.tsx]: @jeyip this is one of the alternatives I had in mind. Does it look more sane from previously?

Complexity is still of the same order (processing all plans), but a smaller surface area of the changes (not touching hooks, etc.). Next step would be to clean this up into a separate hook so it doesn't live like this in the component.

**Comment by jeyip** [comment]
Hey @chriskmnds thanks for pushing changes. I'm being pulled away to take a look at https://github.com/Automattic/wp-calypso/issues/96936#issuecomment-2515446969, but I'll take a look at this PR shortly

File: client/my-sites/plans-features-main/index.tsx (hunk 1 above)
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx]: > this is one of the alternatives I had in mind. Does it look more sane from previously?

From where it was before, the smaller surface area is absolutely better 👍

Currently reading the code to gauge my understanding.

File: client/my-sites/plans-features-main/index.tsx [hunk 2]
```diff
@@ -754,6 +757,52 @@ const PlansFeaturesMain = ( {
 		</div>
 	);
 
+	const usePlanSlugsForAllDisplayedIntervals = () => {
```
Comment (jeyip) [inline client/my-sites/plans-features-main/index.tsx]: `Non-blocking / Nitpick`:

I'm seeing that we have two conventional ways of describing the length of a plan. Just to confirm my understanding, is the difference between "Displayed Interval" and "term" that the former describes a UI concern while the other describes a business concern?

File: client/my-sites/plans-features-main/index.tsx (hunk 1 above)
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx]: > Currently reading the code to gauge my understanding.

Yes this is significantly more easy to follow 🎉 

> Next step would be to clean this up into a separate hook so it doesn't live like this in the component.

Totally agree on encapsulating this complexity.

I suspect you might've already thought about this, but it also seems like an eventual refactor ( doesn't have to be done immediately imo ) would be for the new hook to handle `isAnyPlanPriceDiscounted` logic in the `Header-Price` component.

https://github.com/Automattic/wp-calypso/blob/63d3c4a4da3e8320e71962c64534562006cc2ab3/packages/plans-grid-next/src/components/shared/header-price/index.tsx#L73-L97

I understand that they are distinct concepts, but it seems like we could generate info for `isAnyPlanPriceDiscounted` while we calculate `isAnyGridPlanDiscounted`. Curious about your thoughts though.

File: client/my-sites/plans-features-main/index.tsx (hunk 2 above)
Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: I think one refers to the intervals in the toggle and the URL parameters, the other to the plan term/period.

File: client/my-sites/plans-features-main/index.tsx (hunk 1 above)
Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: Ok - so I moved it now to a separate hook.

> I suspect you might've already thought about this, but it also seems like an eventual refactor ( doesn't have to be done immediately imo ) would be for the new hook to handle isAnyPlanPriceDiscounted logic in the Header-Price component.

Indeed, I had the same thinking. However, we can maybe excuse the two co-existing in that they serve different purposes - one being a header UI concern (that the header price has indeed rendered a discounted price), the other being a general flag into the system/app to enable/disable other things. I think a good refactor might be to turn it into a hook in the plans data-store. I will explore that in a follow-up.

**Comment by chriskmnds** [comment]
@jeyip I think we are ready here to test/deploy. I just can't figure out the type error in the unit tests, where it comes from 🤔

**Comment by jeyip** [comment]
Testing this PR shortly

**Comment by jeyip** [comment]
## Testing
### Requirements

One key concern for us was performance. I ran non-scientific performance evaluations with google chrome dev tools. Note that these metrics will be slower due to the nature of executing code simultaneously with chrome's inspector.

- No CPU or network throttling
- I used the mouse to switch from 1 year annual plans to 2 year plans in the term dropdown selector

#### Trunk
Screenshot 2024-12-04 at 15 03 32

#### This PR
Screenshot 2024-12-04 at 15 02 14

What we're focused on is the scripting time in yellow, and there was a modest 7% increase in time to completion.

For each of the flows listed below:
- Go to /start/plans?flags=plans/term-savings-price-display. Ensure that term savings are displayed for 1y, 2y, and 3y plans.
- Now visit /start/plans?flags=plans/term-savings-price-display&coupon=federate25. Ensure that no term savings are displayed at all. Only the coupon savings. See [here](https://github.com/Automattic/wp-calypso/pull/96631#issuecomment-2512963105) for visuals.

Flows:
- [ ] /start
- [ ] /setup/newsletter

### Browsers
- [x] Chrome

File: client/my-sites/plans-features-main/index.tsx (hunk 1 above)
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx]: > Ok - so I moved it now to a separate hook.

Wonderful! Now we can also explore any necessary performance optimizations in one place ( if necessary ).

**Review by jeyip — APPROVED** [review]
The term dropdown interactions still seem reasonably responsive, and the changes look good to me 👍 I found some strange behavior with currency based special offers mentioned in my testing notes, but that's also present in production.

**Comment by chriskmnds** [comment]
Thanks @jeyip . I will do deploy now.

> I found some strange behavior with currency-based special offers mentioned in my testing notes, but that's also present in production.

Can you port that comment into an issue so we can address it? Feel free to assign me to it. :-)

File: client/my-sites/plans-features-main/index.tsx (hunk 1 above)
Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: Awesome. I'll create an issue for the above in a bit.

Comment (chriskmnds) [reply client/my-sites/plans-features-main/index.tsx]: ^ https://github.com/Automattic/wp-calypso/issues/97132

**Comment by jeyip** [comment]
> Can you port that comment into an issue so we can address it? Feel free to assign me to it. :-)

Of course -- I'm on it 🙂

**Pull request merged**